From: Benjamin Drung <benjamin.drung@canonical.com>
Date: Thu, 22 Sep 2022 19:52:43 +0200
Subject: Fix rpath for private libraries on Linux

Installing audacity on Linux will produce private libraries that have
`RUNPATH` set to the build directory instead of the installation
directory.

The root cause is that the library directory is copied to the
installation directory without touching the libraries. The cmake wiki
says in RPATH handling caveats [1]: "Since install-side RPATH tweaking
is an operation that is done by target-specific installation handling,
any target that should have its install RPATH changed (e.g. to
`CMAKE_INSTALL_RPATH`) needs to end up in the installation via an
`install(TARGETS ...)` signature and not via directory-based copying."

So replace `install(DIRECTORY ...)` by individual `install(TARGETS ...)`
for the libraries and modules. Then cmake will replace the `RUNPATH` to
`$ORIGIN/../lib/audacity`, which is still incorrect. Therefore set
`INSTALL_RPATH` explicitly.

Fixes: https://github.com/audacity/audacity/issues/3289
Forwarded: https://github.com/audacity/audacity/pull/3671
[1] https://gitlab.kitware.com/cmake/community/-/wikis/doc/cmake/RPATH-handling#caveats
Signed-off-by: Benjamin Drung <benjamin.drung@canonical.com>
---
 CMakeLists.txt                                      |  1 +
 cmake-proxies/cmake-modules/AudacityFunctions.cmake | 18 ++++++++++++++++++
 modules/scripting/mod-script-pipe/CMakeLists.txt    |  1 +
 src/CMakeLists.txt                                  |  5 -----
 4 files changed, 20 insertions(+), 5 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2e91683..2e838c3 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -431,6 +431,7 @@ set( INSTALL_PREFIX "${CMAKE_INSTALL_PREFIX}" )
 set( _LIBDIR "${CMAKE_INSTALL_LIBDIR}" )
 set( _DATADIR "${CMAKE_INSTALL_DATADIR}" )
 set( _PKGLIB "${_LIBDIR}/audacity" )
+set( _PKGMODULE "${_PKGLIB}/modules" )
 set( _PKGDATA "${_DATADIR}/audacity/" )
 set( _MANDIR "${CMAKE_INSTALL_MANDIR}" )
 set( _MODDIR "${_PKGLIB}/modules" )
diff --git a/cmake-proxies/cmake-modules/AudacityFunctions.cmake b/cmake-proxies/cmake-modules/AudacityFunctions.cmake
index dee6dcd..6c76c1b 100644
--- a/cmake-proxies/cmake-modules/AudacityFunctions.cmake
+++ b/cmake-proxies/cmake-modules/AudacityFunctions.cmake
@@ -470,6 +470,17 @@ function ( make_interface_alias TARGET REAL_LIBTYTPE )
    endif()
 endfunction()
 
+# Call install(TARGETS...) only on Linux systems (i.e. exclude MacOS and Windows)
+macro( install_target_linux target destination )
+   if( NOT "${CMAKE_GENERATOR}" MATCHES "Xcode|Visual Studio*" AND NOT CMAKE_SYSTEM_NAME MATCHES "Darwin" )
+      install( TARGETS "${target}" DESTINATION "${destination}" )
+   endif()
+endmacro()
+
+macro( install_audacity_module target )
+   install_target_linux( "${target}" "${_PKGMODULE}" )
+endmacro()
+
 function( audacity_module_fn NAME SOURCES IMPORT_TARGETS
    ADDITIONAL_DEFINES ADDITIONAL_LIBRARIES LIBTYPE )
 
@@ -513,6 +524,7 @@ function( audacity_module_fn NAME SOURCES IMPORT_TARGETS
          PROPERTIES
             PREFIX ""
             FOLDER "modules" # for IDE organization
+            INSTALL_RPATH "$ORIGIN/.."
       )
 
       if( NOT CMAKE_SYSTEM_NAME MATCHES "Windows|Darwin" )
@@ -597,6 +609,12 @@ function( audacity_module_fn NAME SOURCES IMPORT_TARGETS
    collect_edges( ${TARGET} "${IMPORT_TARGETS}" ${LIBTYPE} )
    set( GRAPH_EDGES "${GRAPH_EDGES}" PARENT_SCOPE )
 
+   # Note: Some modules set EXCLUDE_FROM_ALL afterwards to not be installed.
+   # Therefore only install libraries, but not modules here.
+   if( NOT REAL_LIBTYPE STREQUAL "MODULE" )
+      install_target_linux( "${TARGET}" "${_PKGLIB}" )
+   endif()
+
    # collect unit test targets if they are present
    if(EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/tests")
       add_subdirectory(tests)
diff --git a/modules/scripting/mod-script-pipe/CMakeLists.txt b/modules/scripting/mod-script-pipe/CMakeLists.txt
index 64cfdd8..a892d95 100644
--- a/modules/scripting/mod-script-pipe/CMakeLists.txt
+++ b/modules/scripting/mod-script-pipe/CMakeLists.txt
@@ -24,3 +24,4 @@ set( LIBRARIES
 )
 audacity_module( mod-script-pipe "${SOURCES}" "${LIBRARIES}"
    "${DEFINES}" "" )
+install_audacity_module( mod-script-pipe )
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 6df0c3e..cc13e7d 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -1412,11 +1412,6 @@ else()
                RUNTIME
                RESOURCE DESTINATION "${_PKGDATA}" )
 
-      install( DIRECTORY "${_DEST}/${_LIBDIR}/"
-               DESTINATION "${_LIBDIR}"
-               USE_SOURCE_PERMISSIONS
-               FILES_MATCHING PATTERN "*.so*" )
-
       install( FILES "${_INTDIR}/audacity.desktop"
                DESTINATION "${_DATADIR}/applications" )
 
